### PR TITLE
Add new npm / glup tasks: `clone:default_plugins`, `build:dev_plugins`, `watch:plugins`

### PR DESCRIPTION
**Pull request by Raruto**
Supersedes: https://github.com/g3w-suite/g3w-client/pull/301 and https://github.com/g3w-suite/g3w-client/pull/262

---

## What's changes:

1. **Add new task** `gulp clone:default_plugins`,
 
    in order to make sure that core client plugins are there before starting develop:

    - [submodule "src/plugins/qplotly"]     <-- https://github.com/g3w-suite/g3w-client-plugin-qplotly.git
    - [submodule "src/plugins/openrouteservice"] <-- https://github.com/g3w-suite/g3w-client-plugin-openrouteservice.git
    - [submodule "src/plugins/qtimeseries"] <-- https://github.com/g3w-suite/g3w-client-plugin-qtimeseries.git
    - [submodule "src/plugins/editing"]     <-- https://github.com/g3w-suite/g3w-client-plugin-editing.git

3. **Add new task** `gulp build:dev_plugins`,
  
   in order to maker sure that core and custom client bundles are up-to-date before starting the `npm run dev` server
   (NB: for performance reasons this script just `build` all plugins once, for watching them all see next point)

4. **Add new script** `npm run watch:plugins`,

   in order to avoid performing the `gulp watch` task on each plugin located in the local `src/plugins/*` directory

7. **Update `README.md`** plugin development instructions

   > ## How to develop
   > You can start the built-in development servers by using the following:
   > ```diff
   > npm run docker:up      # backend server (g3w-admin)
   > npm run dev            # frontend server (g3w-client)
   > + npm run watch:plugins  # watch built-in plugins + any custom plugin (eg. base, eleprofile, sidebar)
   > ```

8. **Remove useful gulp tasks**

    definitely deprecated (and undocumented), the following tasks crash if run manually because their dependent tasks were never implemented:

      - `gulp.task('g3w-admin-client:static', ['build:static']);`
      - `gulp.task('g3w-admin-client:template', ['build:templates']);`
      - `gulp.task('g3w-admin:static', ['build:static']);`
      - `gulp.task('g3w-admin:templates', ['build:templates']);`

## Advantages of this type of solution

1. no headaches with git submodules yet which would require additional documentation on how to clone and update submodules (ref: https://github.com/g3w-suite/g3w-client/pull/301)
2. starts sort of automating the import of core plugins in anticipation of converting them to npm modules (ref: https://github.com/g3w-suite/g3w-client/issues/290)

## Drawbacks

1. ~~The following files are not yet generated automatically (to do this you still need to manually run the `build:plugins` task):~~

   <strike>
   <pre>
    src/plugins/editing/plugin.js
    src/plugins/qplotly/plugin.js
    src/plugins/qtimeseries/plugin.js
   </pre>
   </strike>

2. The version of the cloned repositories is not fixed to a specific release (ie. to a specific branch or commit), so an unsuspecting user could clone a version of the client plugins not immediately compatible with the current version of g3w-client

4. each developer has to worry about keeping various plugin repositories up-to-date locally

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -321,6 +321,29 @@ gulp.task('browser-sync', function() {
...
+  const { execSync } = require('child_process');
+  return new Promise(async done => {
+    for (const pluginName of ['editing', 'qplotly', 'qtimeseries']) {
+      if (!fs.existsSync(`${g3w.pluginsFolder}/${pluginName}/.git`)) {
+        execSync(`git clone https://github.com/g3w-suite/g3w-client-plugin-${pluginName}.git ${g3w.pluginsFolder}/${pluginName}`);
+      }
+      /** @TODO understand why the terminal hangs here (now you must run this 3 times to make it work) */
+      // if (!fs.existsSync(`${g3w.pluginsFolder}/${pluginName}/plugin.js`)) {
+      //   execSync(`npx gulp --gulpfile ${g3w.pluginsFolder}/${pluginName}/gulpfile.js browserify`);  
+      // }
```
Comment (Raruto) [inline gulpfile.js]: Probably somehow related to:

- the implicit directory change (cwd) due to the `--gulpfile` parameter
- the use of the `existsSync` command which does not throws any errors?

Comment (Raruto) [reply gulpfile.js]: Solved by: https://github.com/g3w-suite/g3w-client/pull/309/commits/5f3858de96d328607650d174bc4b81d252df02c2

File: gulpfile.js [hunk 2]
```diff
@@ -336,9 +336,9 @@ gulp.task('clone:default_plugins', function() {
         execSync(`git clone https://github.com/g3w-suite/g3w-client-plugin-${pluginName}.git ${g3w.pluginsFolder}/${pluginName}`);
       }
       /** @TODO understand why the terminal hangs here (now you must run this 3 times to make it work) */
-      // if (!fs.existsSync(`${g3w.pluginsFolder}/${pluginName}/plugin.js`)) {
-      //   execSync(`npx gulp --gulpfile ${g3w.pluginsFolder}/${pluginName}/gulpfile.js browserify`);  
-      // }
+      if (!fs.existsSync(`${g3w.pluginsFolder}/${pluginName}/plugin.js`)) {
+        execSync(`gulp --gulpfile ${g3w.pluginsFolder}/${pluginName}/gulpfile.js --task watch`);
```
Comment (Raruto) [inline gulpfile.js]: `watch` instead of `browserify`?

Comment (volterra79) [reply gulpfile.js]: @Raruto every plugin use:

`gulp.task('watch',['browserify']);
`

From v.4 version it could be better rename it as dev

**Comment by volterra79** [comment]
@Raruto it could be useful add a way here:

https://github.com/g3w-suite/g3w-client/blob/71e12e2c1711a6c58c4f564ff995b4632ddc3ff6/gulpfile.js#L439

to add a task to use plugin in development too

File: gulpfile.js (hunk 2 above)
Comment (Raruto) [reply gulpfile.js]: Note to the reader, each gulpfile also defines a [default task](https://gulpjs.com/docs/en/getting-started/quick-start/#result) that is executed if no task is specified.

gulp default

**For additional information:**

- [g3w-client-plugin-editing/gulpfile.js](https://github.com/g3w-suite/g3w-client-plugin-editing/blob/6b2a25eb7baf6d6d05d0032a1cf908b2be946fb3/gulpfile.js#L86)
- [g3w-client-plugin-qplotly/gulpfile.js](https://github.com/g3w-suite/g3w-client-plugin-qplotly/blob/bb4687f75765b91736e3684055e4284a59ac703e/gulpfile.js#L86)
- [g3w-client-plugin-qtimeseries/gulpfile.js](https://github.com/g3w-suite/g3w-client-plugin-qtimeseries/blob/94c8c81ab007af7f405628660540f72498963804/gulpfile.js#L86)

File: gulpfile.js [hunk 3]
```diff
@@ -45,9 +47,27 @@ const g3w         = require('./config');
...
+
 // Retrieve project dependencies ("g3w-client")
 const dependencies = Object.keys(packageJSON.dependencies).filter(dep => dep !== 'vue');
 
+// Built-in client plugins
+const default_plugins = [
+  'editing',
+  'qplotly',
+  'qtimeseries'
+];
```
Comment (Raruto) [inline gulpfile.js]: @volterra79 should I also include client `openrouteservice` as built-in plugin?

**Ref:**

- [g3w-admin/openrouteservice](https://github.com/g3w-suite/g3w-admin/blob/5e4e294a91bd24fe4cfa0e834d067c7016c64cf3/g3w-admin/openrouteservice)
- [g3w-client-plugin-openrouteservice](https://github.com/g3w-suite/g3w-client-plugin-openrouteservice)

**Comment by Raruto** [comment]
@volterra79 are we ready to merge this?, otherwise we will never test it enough before the next release..

**Review by volterra79 — APPROVED** [review]
@Raruto let's do it

**Pull request merged**